Adding Physical Port Mapping

ittos opened the pull request:
phpIPAM has been a great asset to our business for documenting our network. One thing we needed that wasn't built in was a way to document switch port assignments. These changes enable the user to create port maps, assign them to device, and point specific ports on each map to other devices

ittos committed:
Port Map Permissions

Add permissions to port map

ittos committed:
Copy path

Added copy action to facilitate duplicating port maps

ittos committed:
Port Map

Added port map functionality and abaility tie port maps to devices

ittos committed:
Fixed port adding, removed custom function

Fixed issue adding port to port map, removed custom function from class.Admin which was used to update a devices' port map reference

ittos committed:
Merge pull request #2 from ittos/portMap-function

Pull to 1.4

zombah commented:
Can you please describe more details here about your module or create feature request issue for this. Is it replacement for Physical Circuits or it is possible to assing phys. circuit to port and then assing port to device with your module?
Imho phys. circuits currently connects to devices without ports, so ports module is demanded feature for device1->port1->phys_circuit1->port1->device2 kind of maping.

ittos commented:
It _could_ be modified to point to a physical circuit but this module is a documentation tool to assist in keeping track of switch port assignments. I wanted to keep it simple, intuitive, and informative without a lot of configuration. If this were to need a physical circuit reference it would complicate setup. For a switch serving workstations in an office building it wouldn't make much sense to define a circuit for each connection. Instead, just go to the switch and update port 23 to reference another device. 
 
![PortMap](https://user-images.githubusercontent.com/9108381/61352351-05225000-a822-11e9-8433-70741acc513c.png)

Currently you can only get a device_current, port3->device_other view (With a link to the other device to view the ports from it's perspective) but I'm planning on adding a reverse lookup so you can get a full view of device_current, port3->device_other, port6

ittos committed:
Bug fixes

kustav123 commented:
I followed your pull request and did all the changes is file but after that am getting 500 error please help ..
is there any way to check the error log?

kustav123 commented:
Please note I changed all of this in my running server..

ittos commented:
What version of phpIPAM were you on before starting to add my changes? This is built from the latest v1.4 release

ittos commented on functions/classes/class.Common.php:
```diff
@@ -944,6 +944,7 @@ private function get_valid_actions () {
 		return array(
 		        "add",
 		        "all-add",
+                        "copy".
```
Replace period with comma...

kustav123 replied:
private function get_valid_actions () {
		return array(
		        "add",
		        "all,add",
                        "copy".
		        "edit",
		        "all,edit",
  

??

ittos replied:
This shouldn't be giving you a 500 error but is a small mistake in my typing. The period here would cause a valid action to be "copyedit". Trying to edit anything would return an Invalid Action error

ittos replied:
For your specific issue, my only suggestion would be to undo the changes and verify you're on v1.4

kustav123 replied:
my changes is ok?? actually am not php expert .
 return array(
"add",
"all,add",
"copy".
"edit",
"all,edit",




is that ok?

kustav123 replied:
ok let me undo

kustav123 replied:
phpIPAM IP address management [v1.4] dbversion25

kustav123 replied:
after undoing all changes that I found

kustav123 replied:
will you mind to access via anydesk ?

ittos commented on app/tools/portMaps/single-port-map.php:
```diff
@@ -0,0 +1,164 @@
+            print "<td class='actions'>";
+            $links = [];
+            $links[] = ["type" => "header", "text" => "Manage port"];
+            $links[] = ["type" => "link", "text" => "Edit port", "href" => "", "class" => "open_popup", "dataparams" => " data-script='app/admin/ports/edit.php' data-class='500' data-action='edit' data-id='$port->id'", "icon" => "pencil"];
```
Add data-map_id='$port->map_id' to link.

kustav123 commented:
Hear all my changes please help me.. 
[https://we.tl/t-3G626RnGBs](url)

kustav123 commented:
or can you share the entire frontend as zip or tar ? So that I can just replace my entire frontend

ittos commented:
[phpIPAM1.4+portMap.zip](https://github.com/phpipam/phpipam/files/3402876/phpIPAM1.4%2BportMap.zip)
Here you go

Remember to make a backup of your directory before making any changes

kustav123 commented:
Thanks, friend it worked, it is very very useful for network admin for us. 
Can you tell me what actually is the remote port column of ports table? how to use it on the front end.?


Please have a look on my pull request I believe you can make perfect.

kustav123 commented:
one small issue I found. After creating port map , the port map id is not updating into device table port field.  if I update it from backend it is working fine.

ittos commented on app/admin/portMaps/edit-result.php:
```diff
@@ -0,0 +1,118 @@
+} else {
+    if ($_POST['action'] == "copy") { //If copying a port map also copy ports
+        $ports = $Tools->fetch_multiple_objects("ports", "map_id", $_POST['id']); // Get all ports associated with old map
+        $values["id"] = $Admin->lastId; //Get ID of new port map
```
Move line 92 ($values["id"] = $Admin->lastID;) to before start of if statement on line 90

ittos replied:
See below
https://github.com/ittos/phpipam/commit/d326eb4819243e3c07eb04b88d2fd9faec8aaecd

kustav123 replied:
suddenly am getting "Provided http error code is: 400: Bad Request". when I click on port map menu under device management. 
I changed nothing. even I restored the fresh frontend and recreate the ports and portMaps table

ittos replied:
In order to prevent spam on this section, can we move the conversation to the linked commit on my fork?

kustav123 replied:
ya sure

kustav123 replied:
Please help me about Provided http error code is: 400: Bad Request error

boedyirh replied:
> Move line 92 ($values["id"] = $Admin->lastID;) to before start of if statement on line 90

I think that this will break  function edit PortMap to device

ittos committed:
Move call to get last inserted ID

Fixed adding a port map directly to a device upon creation of the map

ittos commented:
> Thanks, friend it worked, it is very very useful for network admin for us.
> Can you tell me what actually is the remote port column of ports table? how to use it on the front end.?
> 
The remote port was a reference to another port object in order to show the full picture. For example "Port 2 on Switch 1 is plugged into Port 4 on Router 1" but it was too clunky of an implementation. I'm working on a better way of communicating the same information

ittos committed:
Merge pull request #3 from ittos/portMap-function

Port map function

kustav123 commented:
@ittos 
[devices.zip](https://github.com/phpipam/phpipam/files/3407200/devices.zip)
[phpipam.zip](https://github.com/phpipam/phpipam/files/3407206/phpipam.zip)

I have uploaded my entire folder also for your reference.

ittos commented:
> @ittos
> [devices.zip](https://github.com/phpipam/phpipam/files/3407200/devices.zip)
> [phpipam.zip](https://github.com/phpipam/phpipam/files/3407206/phpipam.zip)
> 
> I have uploaded my entire folder also for your reference.

Please reply here:
https://github.com/ittos/phpipam/commit/d326eb4819243e3c07eb04b88d2fd9faec8aaecd

ittos committed:
Fixed 400 error

Added portMaps to admin-menu-config to fix 400 error

ittos committed:
Added Missing types

Adjusted types in SCHEMA file to allow for all inputs in port drop down.
Removed input filter in single port map view

Monkey-Bytes commented:
@ittos Thanks for creating this feature. Currently I have performed a git fetch with your latest head. Looking at the gui I am getting a database error where the table .portMaps doesn't exist. any idea how this may be? Do I need to manually add the table?

ittos commented:
> @ittos Thanks for creating this feature. Currently I have performed a git fetch with your latest head. Looking at the gui I am getting a database error where the table .portMaps doesn't exist. any idea how this may be? Do I need to manually add the table?

@cromulon-actual When logged in as an administrator, go to the verify database section. It should detect that the tables are not in the database and then create them

Monkey-Bytes commented:
@ittos This is the error I'm seeing, with no ability to resolve it.

![image](https://user-images.githubusercontent.com/60146976/73746501-53962e00-4723-11ea-86aa-9c968b563ce8.png)

JeffAlyanak commented:
@ittos, thanks for adding this sorely needed feature! The patch works perfectly for me, hopefully it's merged into the repo soon.

The only thing missing from PHPIPAM now that I truly need is for SNMP autodiscovery to work with this new physical port mapping.

pmehrken commented:
@ittos, I am setting this up and have it installed, however when I click the button to add the port map, nothing happens. What am I missing?

mowfax commented:
Is this compatible with 1.4.2? What are the chances of getting this merged into the master? We're having an overhaul of our network cabling in a few weeks and if there's a chance to get this into phpipam instead of excel that would be really nice.

MartinR2D2 commented:
Is this functionality going to be in any further developments as this would be an extremely useful feature.

gianlucagiacometti commented:
I agree with @zombah .
This functionality is a very nice add-on, but it makes much more sense if integrated with physical circuits.

bitking commented:
I hope something like this would be implemented in the main phpipam code.
And i personally need to be able to map physical circuits (wall outlets in my case) to a switch port number.

kektuto commented:
Hi everyone, I'm new to this and I was wonderig if this feature will be implemented inside the physical circuit or as option for a device. Could anyone need some help with this?
